### PR TITLE
Added new package libidn2 as explicit dependency for gnutls...

### PR DESCRIPTION
**Pull request by Zehvogel**
...also added gnutls as explicit dependency of cmake

I had trouble building cmake on my machine as it tried to use my global libgnutls which was apparently too new. After adding gnutls as an explicit dependency to cmake, I ran into trouble building gnutls as it tried to use my global libidn2 which also was too new. This fixed it.

### CODE REVIEW

File: var/spack/repos/builtin/packages/libidn2/package.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#
+# See the Spack documentation for more information on packaging.
+# ----------------------------------------------------------------------------
+
+from spack import *
+
+
+class Libidn2(AutotoolsPackage):
+    """Libidn2 is a free software implementation of IDNA2008, Punycode and
+    TR46. Its purpose is to encode and decode internationalized domain names."""
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/libidn2/package.py]: @Zehvogel there is a flake error here

File: var/spack/repos/builtin/packages/libidn2/package.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+# ----------------------------------------------------------------------------
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/libidn2/package.py]: @Zehvogel the tests will fail until the boilerplate comment is stripped

Comment (Zehvogel) [reply var/spack/repos/builtin/packages/libidn2/package.py]: @scheibelp my bad, sorry, wonder how I missed that one. I hope everything looks ok now :)

File: var/spack/repos/builtin/packages/libidn2/package.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+    homepage = "https://gitlab.com/libidn/libidn2"
+    url      = "https://ftp.gnu.org/gnu/libidn/libidn2-2.0.5.tar.gz"
+
+    version('2.1.1a', sha256='57666bcf6ecf54230d7bac95c392379561954b57a673903aed4d3336b3048b72')
+    version('2.1.1',  sha256='95416080329298a13269e13175041b530cec3d98b54cafae9424b8dfd22078b1')
+    version('2.1.0',  sha256='032398dbaa9537af43f51a8d94e967e3718848547b1b2a4eb3138b20cad11d32')
+    version('2.0.5',  sha256='53f69170886f1fa6fa5b332439c7a77a7d22626a82ef17e2c1224858bb4ca2b8')
+
+    # FIXME: Add dependencies if required.
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/libidn2/package.py]: @Zehvogel there is one remaining boilerplate comment that is causing the tests to fail.

Comment (Zehvogel) [reply var/spack/repos/builtin/packages/libidn2/package.py]: @scheibelp embarrassing that I missed that one, the tests pass now.

**Comment by scheibelp** [comment]
I'm also curious what system you are building on and if it's possible that this dependency is only required on some systems (since generally up to now I haven't had trouble building/using CMake with Spack for example on RHEL6/7 and Ubuntu).

**Comment by Zehvogel** [comment]
I probably should have gone a bit more into detail in the first post. I opened this PR in a rush..

I am building it on Manjaro and that is part of the issue. You probably didn't run in to any trouble on RHEL and Ubuntu as most of the testing is done there anyway and the package update policies of these distributions are rather "conservative". :-)
I never ran into trouble with spack and cmake in the last two years on the same system as now.

In detail the errors I occured are as follows:
When building cmake (any recent version between 3.12-3.14 haven't tested older ones) with a simple `spack install cmake` the build fails with this message:
```
     808    [ 79%] Linking CXX static library libCMakeServerLib.a
  >> 809    /usr/bin/ld: /usr/lib/libgnutls.so.30: undefined reference to `nettle_rsa_sec_decrypt@@HOGWEED_4'
  >> 810    collect2: error: ld returned 1 exit status
  >> 811    make[2]: *** [Source/CMakeFiles/ccmake.dir/build.make:276: bin/ccmake] Error 1
  >> 812    make[1]: *** [CMakeFiles/Makefile2:1668: Source/CMakeFiles/ccmake.dir/all] Error 2
```
which reduces down to:

- `ld` tries to link against my global `libgnutls` because gnutls is not marked as a dependency in the cmake spack package
- my global `libgnutls` is incompatible to the `nettle` version installed by spack because cmake depends on libarchive and libarchive depends on nettle

In particular, I have `gnutls@3.7.1` installed in `/usr/lib`. My first instinct was to fix the cmake package by including gnutls as an explicit dependency, as the whole reproducible builds purpose of spack pretty much depends on keeping track of all the dependencies a package has. As I am writing this I am retracing my steps and found that [updating nettle to version 3.4.1 should fix the error](https://bugzilla.redhat.com/show_bug.cgi?id=1665484) . However, in my opinion, not having gnutls as explicit dependency is still an issue worth fixing.

After adding gnutls as a dependency I occured the next error during `spack install gnutls`.
```
  >> 3248    /usr/bin/ld: ../lib/.libs/libgnutls.so: undefined reference to `_idn2_punycode_decode'
  >> 3249    collect2: error: ld returned 1 exit status
```
The reason for that is that I have `libidn2@2.1.1` installed globally and [_idn2_punycode_decode' is no longer exposed since version 2.1.0](https://gitlab.com/libidn/libidn2/blob/libidn2-2.1.0/NEWS). Thus, I created a libidn2 package for spack and added it as a dependency to gnutls with proper version handling.

I hope this clarified my intentions with this PR. :-)

**Comment by scheibelp** [comment]
Those are good reasons for including this, I mainly wanted a record of why they were needed if issues are encountered later on other systems (in particular since CMake is a common dependency).

**Comment by mwkrentel** [comment]
@Zehvogel   @scheibelp   @tgamblin 
Could you expand on when and why cmake actually uses gnutls?
Does it depend on +ownlibs or ~ownlibs or ~openssl or something?

I tried building the latest cmake with `--keep-stage` and +openssl,
~openssl, etc.  Every time, I end up with `USE_GNUTLS` turned off.

This is on x86_64, Fedora 26 with gcc 7.3.1 and RPM
gnutls-3.5.18-2.fc26.x86_64.

```
cmake-3.14.4/Utilities/cmcurl/lib/curl_config.h:

/* if GnuTLS is enabled */
/* #undef USE_GNUTLS */
```

Skimming through the cmake source, it seems it has many ways to
satisfy the crypto libs it needs for curl.  `curl_ntlm_core.c` even
says:

```
/* Please keep the SSL backend-specific #if branches in this order:

   1. USE_OPENSSL
   2. USE_GNUTLS_NETTLE
   3. USE_GNUTLS
   4. USE_NSS
   5. USE_MBEDTLS
   6. USE_DARWINSSL
   7. USE_OS400CRYPTO
   8. USE_WIN32_CRYPTO
```

Are there cmake options to diable gnutls and force it to use something
else?  mbedtls is lean, except it's a cmake package.

I care about this (a lot) for two reasons:

(1) gnutls adds 25 dependencies to cmake:

```
gnutls@3.6.8%gcc@7.3.1+guile+zlib arch=linux-fedora26-x86_64
    ^gettext@0.19.8.1%gcc@7.3.1+bzip2+curses+git~libunistring+libxml2 patches=9acdb4e73f67c241b5ef32505c9ddf7cf6884ca8ea661692f21dca28483b04b8 +tar+xz arch=linux-fedora26-x86_64
        ^bzip2@1.0.6%gcc@7.3.1+shared arch=linux-fedora26-x86_64
            ^diffutils@3.7%gcc@7.3.1 arch=linux-fedora26-x86_64
        ^libxml2@2.9.9%gcc@7.3.1~python arch=linux-fedora26-x86_64
            ^libiconv@1.15%gcc@7.3.1 arch=linux-fedora26-x86_64
            ^pkgconf@1.6.0%gcc@7.3.1 arch=linux-fedora26-x86_64
            ^xz@5.2.4%gcc@7.3.1 arch=linux-fedora26-x86_64
            ^zlib@1.2.11%gcc@7.3.1+optimize+pic+shared arch=linux-fedora26-x86_64
        ^ncurses@6.1%gcc@7.3.1~symlinks~termlib arch=linux-fedora26-x86_64
        ^tar@1.31%gcc@7.3.1 arch=linux-fedora26-x86_64
    ^guile@2.2.0%gcc@7.3.1+readline+threads arch=linux-fedora26-x86_64
        ^bdw-gc@8.0.0%gcc@7.3.1+libatomic-ops arch=linux-fedora26-x86_64
            ^libatomic-ops@7.6.6%gcc@7.3.1 arch=linux-fedora26-x86_64
        ^gmp@6.1.2%gcc@7.3.1 arch=linux-fedora26-x86_64
            ^autoconf@2.69%gcc@7.3.1 arch=linux-fedora26-x86_64
                ^m4@1.4.18%gcc@7.3.1 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,c0a408fbffb7255fcc75e26bd8edab116fc81d216bfd18b473668b7739a4158e,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 +sigsegv arch=linux-fedora26-x86_64
                    ^libsigsegv@2.11%gcc@7.3.1 arch=linux-fedora26-x86_64
                ^perl@5.24.4%gcc@7.3.1+cpanm patches=0eac10ed90aeb0459ad8851f88081d439a4e41978e586ec743069e8b059370ac +shared+threads arch=linux-fedora26-x86_64
            ^automake@1.16.1%gcc@7.3.1 arch=linux-fedora26-x86_64
            ^libtool@2.4.6%gcc@7.3.1 arch=linux-fedora26-x86_64
        ^libffi@3.2.1%gcc@7.3.1 arch=linux-fedora26-x86_64
        ^libunistring@0.9.10%gcc@7.3.1 arch=linux-fedora26-x86_64
        ^readline@7.0%gcc@7.3.1 arch=linux-fedora26-x86_64
    ^nettle@3.4.1%gcc@7.3.1 arch=linux-fedora26-x86_64
```

(2) Right now, the build for gnutls falls over in the config test for
guile.  (I had to build with ~guile just to get this far.)

```
1 error found in build log:
     912    configure: checking for guile 2.2
     913    configure: found guile 2.2
     914    checking for guile-2.2... no
     915    checking for guile2.2... no
     916    checking for guile-2... no
     917    checking for guile2... /usr/bin/guile2
  >> 918    configure: error: found development files for Guile 2.2, but /usr/bin/guile2 has effective version 2.0
```

This is near blocking because it's breaking the build for cmake.
So yes, issues are encountered on other systems.

**Comment by mwkrentel** [comment]
@Zehvogel @scheibelp @tgamblin
Would anyone mind if I changed `depends_on('gnutls')` to a variant,
and preferably with default off?

Gnutls is just one way of many to provide the ssl libraries for curl,
openssl is another.  And cmake doesn't use gnutls without the option
`-DUSE_GNUTLS=1` which the spack recipe doesn't add.  It's too bad
that mbedtls is a cmake package, mbedtls is simple and lean and would
also work.

And I'm not saying that libgnutls from /usr/bin is better than the
spack-built version.  Cmake doesn't use either of them without
`-DUSE_GNUTLS=1`.

So, in this case, depends on gnutls adds 25 dependencies that cmake
drops on the floor.  Besides the bloat, it also makes the cmake build
fragile.  There are two pending PRs for broken prereqs in the gnutls
dependency tree: #11607 and #11611.

**Pull request merged**